File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;

bool visited[8];
int res = 0;

int func(vector<vector<int>> dungeons, int k, int cnt)
{
    res = max(res, cnt);
    for(int i=0; i<dungeons.size(); i++)
    {
        // 방문한 던전이거나 최소보다 현재가 작으면 넘어가
        if(visited[i] || k < dungeons[i][0])
            continue;
        visited[i] = true;
        func(dungeons, k - dungeons[i][1], cnt + 1);
        visited[i] = false;
    }
    return res;
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = func(dungeons, k, 0);
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    boolean[] select;
    int num, maxCnt;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        
        // 던전은 최대 8개
        // 최대한 많이
        // 순열
        
        num = dungeons.length;
        select = new boolean[num];
        
        for(int i=0; i<num; i++){
            if(dungeons[i][0] <= k){
                select[i] = true;
                dfs(k-dungeons[i][1], 1, dungeons);
                select[i] = false;
            }
        }
        
        return maxCnt;
    }
    
    public void dfs(int left, int cnt, int[][] dungeons){

        for(int i=0; i<num; i++){
            if(!select[i] && dungeons[i][0] <= left){
                select[i] = true;
                dfs(left-dungeons[i][1], cnt+1, dungeons);
                select[i] = false;
            }
        }
        maxCnt = Math.max(maxCnt, cnt);
    }
}